opm/output/Wells.hpp

```diff
@@ -0,0 +1,173 @@
+            /// Set the value specified by m. Throws an exception if multiple
+            /// values are requested. Returns a self-reference to support
+            /// chaining.
+            inline Rates& set( opt m, double value );
+
+        private:
+            double& get_ref( opt, double& );
+
+            opt mask = static_cast< opt >( 0 );
+
+            double wat;
+            double oil;
+            double gas;
+            double polymer;
+    };
+
+    struct Completion {
+        size_t logical_cartesian_index;
+        Rates rates;
+    };
+
+    struct Well {
+        Rates rates;
+        double bhp;
+        std::map< size_t /*logical_cartesian_index*/, Completion > completions;
+    };
+
+    struct Wells {
+        double step_length;
+        std::map< std::string, Well > wells;
+    };
+
+    /* IMPLEMENTATIONS */
+
+    inline Rates::opt operator|( Rates::opt x, Rates::opt y ) {
+        using sz = Rates::enum_size;
+        return static_cast< Rates::opt >
+            ( static_cast< sz >( x ) | static_cast< sz >( y ) );
+    }
+
+    inline Rates::opt operator&( Rates::opt x, Rates::opt y ) {
+        using sz = Rates::enum_size;
+        return static_cast< Rates::opt >
+            ( static_cast< sz >( x ) & static_cast< sz >( y ) );
+    }
+
+    inline Rates::opt& operator|=( Rates::opt& x, Rates::opt y ) {
+        return x = x | y;
+    }
+
+    inline bool Rates::has( opt m ) {
+        return (this->mask & m) == m;
+    }
+
+    inline double Rates::get( opt m ) {
+        if( !this->has( m ) )
+            throw std::invalid_argument(
+                    "Invalid bitmask "
+                    + std::to_string( static_cast< Rates::enum_size >( m ) )
+                    + ". Uninitialized value." );
+
+        double errval;
+        auto& ret = this->get_ref( m, errval );
+
+        if( std::addressof( ret ) == std::addressof( errval ) )
+            throw std::invalid_argument( "Invalid bitmask." );
+
+        return ret;
+    }
+
+    inline double Rates::get( opt m, double errval ) {
+        if( !this->has( m ) ) return errval;
+
+        return this->get_ref( m, errval );
+    }
+
+    inline Rates& Rates::set( opt m, double value ) {
+        double errval;
+        auto& x = this->get_ref( m, errval );
+
+        if( std::addressof( errval ) == std::addressof( x ) )
+            throw std::invalid_argument( "Invalid bitmask." );
+
+        this->mask |= m;
+        x = value;
+        return *this;
+    }
+
+    /*
+     * To avoid error-prone and repetitve work when extending rates with new
+     * values, the get+set methods use this helper get_ref to determine what
+     * member to manipulate. To add a new option, just add another case
+     * corresponding to the enum entry in Rates to this function.
+     *
+     * This is an implementation detail and understanding this has no
+     * significant impact on correct use of the class.
+     */
```

Comment:
Maybe. It does however ease maintenance by providing *one* place that must be updated for supporting a new variable, instead of 3.